File: app/local_plugins/translate_server/android/src/main/kotlin/com/aiagent/translate_server/TranslateServerPlugin.kt
package com.aiagent.translate_server

import android.content.Context
import android.util.Log
import com.aiagent.plugin_interface.NativeAgentConfig
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.*
import java.util.UUID

/**
 * translate_server Flutter plugin —— Method/Event channel 入口。
 *
 *  - method channel `translate_server/method`:
 *      • startCallTranslation(arg: Map) → returns sessionId
 *      • stopActiveSession()
 *      • activeSessionId() → returns String?
 *  - event channel `translate_server/events`:
 *      字幕 / 状态 / 错误 / 连接状态（详见 [TranslateEvents]）
 *
 * 编排逻辑全部在 [TranslateServerCore] / [CallTranslationSession]，本类只做调度。
 */
class TranslateServerPlugin : FlutterPlugin {

    companion object {
        private const val TAG = "TranslateServerPlugin"
    }

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private val mainScope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private lateinit var appContext: Context

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        appContext = binding.applicationContext

        eventChannel = EventChannel(binding.binaryMessenger, "translate_server/events")
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
                TranslateServerCore.bindEmitter { payload ->
                    mainScope.launch { runCatching { events?.success(payload) } }
                }
            }
            override fun onCancel(arguments: Any?) {
                TranslateServerCore.unbindEmitter()
                eventSink = null
            }
        })

        methodChannel = MethodChannel(binding.binaryMessenger, "translate_server/method")
        methodChannel.setMethodCallHandler { call, result ->
            try {
                handleMethod(call, result)
            } catch (e: Exception) {
                Log.e(TAG, "method ${call.method} failed", e)
                result.error(e::class.java.simpleName, e.message, null)
            }
        }
    }

    private fun handleMethod(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startCallTranslation" -> {
                val args = call.arguments as? Map<*, *>
                    ?: return result.error("InvalidArgument", "arguments must be a map", null)
                val req = parseCallRequest(args)
                val sessionId = (args["sessionId"] as? String)
                    ?: ("ts_${System.currentTimeMillis()}_${UUID.randomUUID().toString().take(6)}")
                val id = TranslateServerCore.startCallTranslation(sessionId, req, appContext)
                result.success(id)
            }

            "stopActiveSession" -> {
                TranslateServerCore.stopActive()
                result.success(null)
            }

            "activeSessionId" -> {
                result.success(TranslateServerCore.activeSessionId())
            }

            "startFaceToFaceTranslation",
            "startAudioTranslation" -> {
                result.error("translate.not_implemented",
                    "${call.method} not implemented yet", null)
            }

            else -> result.notImplemented()
        }
    }

    private fun parseCallRequest(args: Map<*, *>): CallTranslationRequest {
        val uplinkAgentType = args["uplinkAgentType"] as? String
            ?: error("uplinkAgentType required")
        val downlinkAgentType = args["downlinkAgentType"] as? String
            ?: error("downlinkAgentType required")
        val uplinkConfigMap = (args["uplinkConfig"] as? Map<*, *>)
            ?: error("uplinkConfig required")
        val downlinkConfigMap = (args["downlinkConfig"] as? Map<*, *>)
            ?: error("downlinkConfig required")
        val userLanguage = (args["userLanguage"] as? String) ?: ""
        val peerLanguage = (args["peerLanguage"] as? String) ?: ""

        return CallTranslationRequest(
            uplinkAgentType = uplinkAgentType,
            uplinkConfig = NativeAgentConfig.fromMap(uplinkConfigMap),
            downlinkAgentType = downlinkAgentType,
            downlinkConfig = NativeAgentConfig.fromMap(downlinkConfigMap),
            userLanguage = userLanguage,
            peerLanguage = peerLanguage,
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        TranslateServerCore.unbindEmitter()
        TranslateServerCore.stopActive()
        mainScope.cancel()
    }
}

File: app/local_plugins/translate_server/ios/Classes/TranslateServerPlugin.swift
import Flutter
import Foundation
import ai_plugin_interface
import os.log

/// translate_server Flutter plugin（iOS）—— 与 Android `TranslateServerPlugin` 对齐。
public class TranslateServerPlugin: NSObject, FlutterPlugin {

    private static let log = OSLog(subsystem: "com.aiagent.translate_server", category: "Plugin")

    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private let streamHandler = TranslateEventStreamHandler()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = TranslateServerPlugin()
        let messenger = registrar.messenger()

        instance.eventChannel = FlutterEventChannel(
            name: "translate_server/events",
            binaryMessenger: messenger
        )
        instance.eventChannel?.setStreamHandler(instance.streamHandler)

        instance.methodChannel = FlutterMethodChannel(
            name: "translate_server/method",
            binaryMessenger: messenger
        )
        registrar.addMethodCallDelegate(instance, channel: instance.methodChannel!)

        TranslateServerCore.shared.bindContext(messenger: messenger) { payload in
            instance.streamHandler.send(payload)
        }
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        do {
            try dispatch(call, result: result)
        } catch {
            result(FlutterError(code: "TranslateServer", message: error.localizedDescription, details: nil))
        }
    }

    private func dispatch(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        switch call.method {
        case "startCallTranslation":
            guard let args = call.arguments as? [String: Any] else {
                result(FlutterError(code: "InvalidArgument", message: "arguments must be a map", details: nil))
                return
            }
            let req = try parseCallRequest(args)
            let sessionId = (args["sessionId"] as? String)
                ?? "ts_\(Int64(Date().timeIntervalSince1970 * 1000))_\(UUID().uuidString.prefix(6))"
            do {
                let id = try TranslateServerCore.shared.startCallTranslation(sessionId: sessionId, request: req)
                result(id)
            } catch let nsErr as NSError {
                result(FlutterError(
                    code: codeFromMessage(nsErr.localizedDescription, defaultCode: "translate.start_failed"),
                    message: nsErr.localizedDescription,
                    details: nil
                ))
            }

        case "stopActiveSession":
            TranslateServerCore.shared.stopActive()
            result(nil)

        case "activeSessionId":
            result(TranslateServerCore.shared.activeSessionId())

        case "startFaceToFaceTranslation", "startAudioTranslation":
            result(FlutterError(code: "translate.not_implemented",
                                message: "\(call.method) not implemented yet", details: nil))

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func parseCallRequest(_ args: [String: Any]) throws -> CallTranslationRequest {
        guard let uplinkAgentType = args["uplinkAgentType"] as? String, !uplinkAgentType.isEmpty else {
            throw NSError(domain: "TranslateServer", code: -1,
                          userInfo: [NSLocalizedDescriptionKey: "uplinkAgentType required"])
        }
        guard let downlinkAgentType = args["downlinkAgentType"] as? String, !downlinkAgentType.isEmpty else {
            throw NSError(domain: "TranslateServer", code: -1,
                          userInfo: [NSLocalizedDescriptionKey: "downlinkAgentType required"])
        }
        guard let uplinkConfigMap = args["uplinkConfig"] as? [String: Any] else {
            throw NSError(domain: "TranslateServer", code: -1,
                          userInfo: [NSLocalizedDescriptionKey: "uplinkConfig required"])
        }
        guard let downlinkConfigMap = args["downlinkConfig"] as? [String: Any] else {
            throw NSError(domain: "TranslateServer", code: -1,
                          userInfo: [NSLocalizedDescriptionKey: "downlinkConfig required"])
        }
        let userLanguage = (args["userLanguage"] as? String) ?? ""
        let peerLanguage = (args["peerLanguage"] as? String) ?? ""
        return CallTranslationRequest(
            uplinkAgentType: uplinkAgentType,
            uplinkConfig: NativeAgentConfig.fromMap(uplinkConfigMap),
            downlinkAgentType: downlinkAgentType,
            downlinkConfig: NativeAgentConfig.fromMap(downlinkConfigMap),
            userLanguage: userLanguage,
            peerLanguage: peerLanguage
        )
    }

    private func codeFromMessage(_ msg: String, defaultCode: String) -> String {
        let prefix = msg.split(separator: ":").first.map(String.init) ?? ""
        return prefix.hasPrefix("translate.") ? prefix : defaultCode
    }
}

final class TranslateEventStreamHandler: NSObject, FlutterStreamHandler {
    private let lock = NSLock()
    private var sink: FlutterEventSink?

    func onListen(withArguments _: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        lock.lock(); sink = events; lock.unlock()
        return nil
    }

    func onCancel(withArguments _: Any?) -> FlutterError? {
        lock.lock(); sink = nil; lock.unlock()
        return nil
    }

    func send(_ payload: [String: Any?]) {
        let s: FlutterEventSink?
        lock.lock(); s = sink; lock.unlock()
        guard let s = s else { return }
        let cleaned = payload.reduce(into: [String: Any]()) { acc, kv in
            if let v = kv.value { acc[kv.key] = v }
        }
        if Thread.isMainThread {
            s(cleaned)
        } else {
            DispatchQueue.main.async { s(cleaned) }
        }
    }
}
